### DISCUSSION TITLE
Nice header only theadpool!

### DISCUSSION THREAD

**Discussion by nitro44x** [Hoshiningen/ThreadPool#3] [Q&A] [answered]
I see you're putting that concurrency book to good use!

quick question: In the worker thread loop, don't you want to grab a job off the queue, then release the lock BEFORE running the job? Seems like running the job, popping it off the queue, then releasing will serialize everything.

Or..... am I reading that wrong? I'll admit, I usually need to reread docs on the locks and condition variables...

**Comment by Hoshiningen** [comment]
I believe you are correct! I can see why you asked about benchmarks earlier. As it is now, the lock isn't released until after the task has executed. Great catch!

**Reply by nitro44x** [reply]
I may or may not have done something similar at work :)

**Comment by Hoshiningen** [comment, marked as answer]
This has been resolved in the commit: https://github.com/Hoshiningen/ThreadPool/commit/e36b58faab8f6fdc58898a3d01a87fa47b1f7717
